File: src/cn/timelives/java/math/abstractAlgebra/calculator/MonoidCalculator.java
/**
 * 2018-02-27
 */
package cn.timelives.java.math.abstractAlgebra.calculator;

/**
 * A MonoidCalculator is a calculator specialized for monoid.
 * @author liyicheng
 * 2018-02-27 17:40
 *
 */
public interface MonoidCalculator<T> extends SemigroupCalculator<T> {
	
	
	/**
	 * Returns the identity element of the semigroup.
	 * @return
	 */
	public T getIdentity();
	
	
	/**
	 * @param n a non-negative number
	 */
	@Override
	default T gpow(T x, long n) {
		if(n==0) {
			return getIdentity();
		}
		return SemigroupCalculator.super.gpow(x, n);
	}
}

File: src/cn/timelives/java/math/algebra/abstractAlgebra/calculator/MonoidCalculator.kt
/**
 * 2018-02-27
 */
package cn.timelives.java.math.algebra.abstractAlgebra.calculator

/**
 * A MonoidCalculator is a calculator specialized for monoid.
 * @author liyicheng
 * 2018-02-27 17:40
 */
interface MonoidCalculator<T : Any> : SemigroupCalculator<T> {


    /**
     * Returns the identity element of the semigroup.
     * @return
     */
    val identity: T


    /**
     * @param n a non-negative number
     */
    override fun gpow(x: T, n: Long): T {
        return if (n == 0L) {
            identity
        } else super.gpow(x, n)
    }
}
